blockchain/fastsync.go

```diff
@@ -0,0 +1,237 @@
+			break
+		} else if err != nil {
+			log.Errorf("Error reading UTXO set: %s", err.Error())
+			return err
+		}
+		totalRead += n
+
+		// The last four bytes that we read is the length of the script
+		scriptLen = binary.LittleEndian.Uint32(buf52[48:])
+		if scriptLen > maxScriptLen {
+			log.Error("Read invalid UTXO script length", totalRead)
+			return errors.New("invalid script length")
+		}
+
+		// Read the script
+		pkScript = make([]byte, scriptLen)
+		n, err = file.Read(pkScript)
+		if err != nil {
+			log.Errorf("Error reading UTXO set: %s", err.Error())
+			return err
+		}
+		totalRead += n
+
+		serializedUtxo = make([]byte, 52+scriptLen)
+		serializedUtxo = append(buf52, pkScript...)
+
+		jobsChan <- serializedUtxo
+	}
+	close(jobsChan)
+
+	// Read each result and add the returned hash to the
+	// existing multiset.
+	m := bchec.NewMultiset(bchec.S256())
+	for i := 0; i < numWorkers; i++ {
+		result := <-resultsChan
+		if result.err != nil {
+			log.Errorf("Error processing UTXO set: %s", err.Error())
+			return err
+		}
+		m.Merge(result.m)
+	}
+	close(resultsChan)
+
+	if err = b.utxoCache.Flush(FlushRequired, state); err != nil {
+		log.Errorf("Error processing UTXO set: %s", err.Error())
+		return err
+	}
+
+	if err = b.index.flushToDB(); err != nil {
+		log.Errorf("Error processing UTXO set: %s", err.Error())
+		return err
+	}
+
+	utxoHash := m.Hash()
+
+	// Make sure the hash of the UTXO set we downloaded matches the expected hash.
+	if !checkpoint.UtxoSetHash.IsEqual(&utxoHash) {
+		log.Errorf("Downloaded UTXO set hash does not match checkpoint."+
+			" Expected %s, got %s.", checkpoint.UtxoSetHash.String(), m.Hash().String())
+		return AssertError("downloaded invalid UTXO set")
+	}
+
+	log.Infof("Verification complete. UTXO hash %s.", m.Hash().String())
+
+	// Signal fastsync complete
+	close(b.fastSyncDone)
+
+	return nil
+}
+
+// result holds a multiset with a hash of all the UTXOs read by
+// this worker and a possible error.
+type result struct {
+	m   *bchec.Multiset
+	err error
+}
+
+// worker handles the work of deserializing the UTXO, calculating the ECMH hash of
+// each serialized UTXO as well as saving it into the utxoCache. The resulting
+// multiset or an error will be returned over the results chan when the jobs
+// chan is closed.
+func worker(cache *utxoCache, state *BestState, jobs <-chan []byte, results chan<- *result) {
+	var (
+		err      error
+		m        = bchec.NewMultiset(bchec.S256())
+		entry    *UtxoEntry
+		outpoint *wire.OutPoint
+	)
+	for serializedUtxo := range jobs {
+		m.Add(serializedUtxo)
+
+		outpoint, entry, err = deserializeUtxoCommitmentFormat(serializedUtxo)
+		if err != nil {
+			log.Errorf("Error deserializing UTXO: %s", err.Error())
+			results <- &result{err: err}
+			return
+		}
+
+		if err = cache.AddEntry(*outpoint, entry, true); err != nil {
+			results <- &result{err: err}
+			return
+		}
+
+		if err = cache.Flush(FlushIfNeeded, state); err != nil {
```

Comment:
I think the reason this flush is unreliable is because of the following code inside the `flush()` call. Since you are restoring the utxos I don't think the last state changes, which means flushes will always be ignored.

```
	// If we performed a flush in the current best state, we have nothing to do.
	if bestState.Hash == s.lastFlushHash {
		return nil
	}
```

Reply:
Good catch